Code complete nested columns

cccs-jc opened the pull request:
This PR should be applied after the documentation PR I've created earlier.

This pull request contains the main support for code completion of nested columns, it adds support for the following:

- Nested columns names using multi-part dot separated paths
- Support for spaces in nested columns names (backtiks each segment)
- Visually distinct icons for `table`, `column`, `alias`, `functions` suggestions
- Added support for code-completion of
	- Functions including descriptions
	- Table alias
	- Table alias followed by dot
	- Partially typed columns including multi-part fields (needed insertText)
        - Map and Array subscripts `map_col['key'].<tab>`, `array_col[0].<tab>`

More details can be found in example/jupyterlab/README.md

This is a fairly large PR which I've tested as much as I could. Please feel free to reach out if you have any questions and want to discuss changes.

Cheers

Commit:
functions and nested columns code complete

Commit:
added support for catalog
and fully qualified table names

Commit:
fix FromClauseParser
it was not picking up the reservedMap of keywords
this had the side effect of allowing the WHERE
keyword to be an AS identifier

Commit:
refactoring

Commit:
refactor to a Completer class

Commit:
launch files

joe-re commented on packages/server/package.json:
```diff
@@ -36,7 +36,7 @@
   ],
   "dependencies": {
-    "@joe-re/sql-parser": "^0.11.5",
+    "@joe-re/sql-parser": "file:../sql-parser",
```
👍

joe-re commented on packages/server/test/complete.test.ts:
```diff
@@ -15,11 +15,27 @@ describe('keyword completion', () => {
+
+    test("complete FROM word with norm columns", () => {
+      const result = complete('SELECT d, f AS F', { line: 0, column: 16 })
+      // TODO: this is not correct but difficult to fix...
```
Thanks for leaving this as a comment! 🙏

joe-re commented on packages/server/test/complete.test.ts:
```diff
@@ -15,11 +15,27 @@ describe('keyword completion', () => {
+      expect(result.candidates.length).toEqual(1)
+      expect(result.candidates[0].label).toEqual('FROM')
+    })
+    test("complete FROM word with norm columns", () => {
```
```suggestion
    test("complete AS word with norm columns", () => {
```

joe-re commented on packages/server/test/complete.test.ts:
```diff
@@ -83,55 +120,144 @@ describe('keyword completion', () => {
+  // TODO: should support this
+  test("complete ColumnName:cursor on first char:using back tick", () => {
+    const result = complete('SELECT `t.C FROM TABLE1 as t', { line: 0, column: 10 }, SIMPLE_SCHEMA)
+    expect(result.candidates.length).toEqual(0)
```
@cccs-jc 
Can we use `test.skip` and write an expected behavior originally?
https://jestjs.io/docs/next/api#testskipname-fn

joe-re commented on packages/server/test/complete.test.ts:
```diff
@@ -168,113 +294,479 @@ describe('From clause', () => {
+  })
+
+  // TODO should support this
+  test("not complete when a cursor is on dot in back tick column name", () => {
```
@cccs-jc 
Can we use `test.skip` and write an expected behavior originally?
https://jestjs.io/docs/next/api#testskipname-fn

joe-re commented on packages/server/test/complete.test.ts:
```diff
@@ -168,113 +294,479 @@ describe('From clause', () => {
+  expect(result.candidates.length).toEqual(11)
+})
+
+test("conplete columns from alias that start chars same as other table", () => {
```
```suggestion
test("complete columns from alias that start chars same as other table", () => {
```

joe-re commented on packages/server/test/complete.test.ts:
```diff
@@ -168,113 +294,479 @@ describe('From clause', () => {
 }
 
+
+test("conplete columns from alias that start chars same as other table", () => {
```
```suggestion
test("complete columns from alias that start chars same as other table", () => {
```

joe-re commented on packages/server/test/complete.test.ts:
```diff
@@ -403,4 +919,56 @@ describe('DELETE statement', () => {
+    expect(completion.insertText).toEqual('.subcolumn2')
+    expect(completion.filterText).toEqual('.subcolumn2')
+  })
+  test('complete aliased nested comlumn name2', () => {
```
Can we name it something intentionally instead of "2"?

joe-re commented on packages/server/test/complete.test.ts:
```diff
@@ -292,7 +784,7 @@ test("conplete columns from duplicated alias", () => {
 })
 
-test("conplete columns innside function", () => {
+test("conplete columns innside function 1", () => {
```
```suggestion
test("complete columns inside function 1", () => {
```

Can we name it something intentionally instead of "1" - "4"?

cccs-jc replied:
Hey @joe-re, will do thanks for reviewing. I'll try to do this today. Cheers.

Commit:
fix "debouce" issue

when user chooses a completion item
jupyterlab re-triggers code completion with that last text entry
this causes complete to be called again and for the same suggestion to be
made

Commit:
JupyterLab sends invalide completion requeset

after choosing a word like city, jupyter lab issues a new completion for
the trigger character 'y' even though
we specify to call us for the '.' character

defensively returning []

Commit:
code review fixes

Commit:
added timing code

Commit:
fix completion when not triggered by dot

in that case we should not use insertText and findText

Commit:
Added utility to automatically generate all column names
for a given table in a SELECT statement

Commit:
Added utility to automatically generate join conditions on matching columns from other tables

Commit:
documentation

Commit:
refering to code-complete-nested-columns branch
will change this once it's merged into release

Commit:
jupyterLabMode

isolated the difference between vscode and jupyterlab into one place

added a flag in the configuration to run in either jupyterLabMode or not

It looks like this is a bug in JupyterLab. I will report it and possibly eliminate the need for this work around.

Commit:
unused import

joe-re commented:
@cccs-jc 
Thanks for your great work! I'm looking through your changes right now. Please give me more little time, thanks. 🙏

cccs-jc commented:
@joe-re
Take your time. I made a lot of changes.

joe-re commented on packages/server/src/createServer.ts:
```diff
@@ -192,17 +198,23 @@ export function createServerWithConnection(connection: Connection) {
+    let pos = { line: docParams.position.line, column: docParams.position.character }
+    var setting = SettingStore.getInstance().getSetting()
+    const candidates = complete(text, pos, schema, setting.jupyterLabMode).candidates
+    if (logger.isDebugEnabled()) logger.debug('onCompletion returns: ' + JSON.stringify(candidates))
```
Why should this be in if condition?
As long as it uses `logger.debug`, it doesn't need to concern about log level on code lines.

cccs-jc replied:
You are correct that `logger.debug` will get stubbed out when debugging is turned off. However computing the string passed to the `logger.debug` function will still happen. In cases where computing the string to be passed is costly you might want to place a condition before the `logger.debug`.

This is what I do in Java anyways. I assume the same is true in JavaScript.

joe-re commented on packages/sql-parser/index.d.ts:
```diff
@@ -143,6 +143,7 @@ export type FromTableNode = TableNode | SubqueryNode | IncompleteSubqueryNode
 
 export type TableNode = {
   type: 'table',
+  catalog: string,
```
```suggestion
  catalog: string | null,
```

Commit:
code review fix

joe-re commented:
@cccs-jc Thanks for your great work! I'll work on releasing this (also fix [this](https://github.com/joe-re/sql-language-server/issues/80)) in a few days.
Thanks again!

cccs-jc commented:
Thanks for reviewing @joe-re
